linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -715,6 +716,59 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
     fb.buildAndAddToModule()
   }
 
+  private def genUnboxByteOrShort(typeRef: PrimRef)(implicit ctx: WasmContext): Unit = {
+    /* As long as we only support Unchecked asInstanceOfs, the unboxing
+     * functions for Byte and Short actually do exactly the same thing.
+     * We keep them separate so that the rest of the codebase is clearer.
+     */
+
+    val fb = newFunctionBuilder(genFunctionID.unbox(typeRef))
+    val xParam = fb.addParam("x", RefType.anyref)
+    fb.setResultType(Int32)
+
+    // If x is a (ref i31), extract it
+    fb.block(RefType.anyref) { xIsNotI31Label =>
+      fb += LocalGet(xParam)
+      fb += BrOnCastFail(xIsNotI31Label, RefType.anyref, RefType.i31)
+      fb += I31GetS
+      fb += Return
+    }
+
+    // Otherwise, it must be null, so return 0
+    fb += Drop
+    fb += I32Const(0)
+
+    fb.buildAndAddToModule()
+  }
+
+  private def genTestByteOrShort(typeRef: PrimRef, signExtend: SimpleInstr)(
+        implicit ctx: WasmContext): Unit = {
+
+    val fb = newFunctionBuilder(genFunctionID.typeTest(typeRef))
+    val xParam = fb.addParam("x", RefType.anyref)
+    fb.setResultType(Int32)
+
+    val intValueLocal = fb.addLocal("intValue", Int32)
+
+    // If x is a (ref i31), extract it and that it sign-extends to itself
+    fb.block(RefType.anyref) { xIsNotI31Label =>
+      fb += LocalGet(xParam)
+      fb += BrOnCastFail(xIsNotI31Label, RefType.anyref, RefType.i31)
+      fb += I31GetS
+      fb += LocalTee(intValueLocal)
+      fb += LocalGet(intValueLocal)
+      fb += signExtend
+      fb += I32Eq
+      fb += Return
+    }
+
+    // Otherwise, return false
+    fb += Drop
+    fb += I32Const(0)
```

Comment:
Don't we have to allow for a slow path as well if someone passes a number created in JS?